src/main/java/in/org/projecteka/hiu/consent/ConsentService.java

```diff
@@ -346,71 +354,64 @@ private String getCmSuffix(Consent consent) {
+        if (consentArtefactResponse.getError() != null) {
+            logger.error(String.format("[ConsentService] Received error response for consent-artefact. HIU RequestId=%s, Error code = %d, message=%s",
+                    consentArtefactResponse.getResp().getRequestId(),
+                    consentArtefactResponse.getError().getCode(),
+                    consentArtefactResponse.getError().getMessage()));
+            return Mono.empty();
         }
 
+        if (consentArtefactResponse.getConsentArtefactResponse() != null) {
+            this.gatewayResponseCache.put(consentArtefactResponse.getResp().getRequestId(), Optional.of(consentArtefactResponse.getConsentArtefactResponse()));
+        }
+
+        return Mono.empty();
+    }
+
+    private Mono<Void> processConsentNotification(ConsentNotification notification, LocalDateTime localDateTime) {
+        var consentTask = consentTasks.get(notification.getStatus());
+        if (consentTask == null) {
+            return Mono.error(ClientError.validationFailed());
+        }
+        return consentTask.perform(notification, localDateTime);
     }
 
     @PostConstruct
     private void postConstruct() {
         consentTasks.put(GRANTED, new GrantedConsentTask(
-                    gatewayServiceClient,centralRegistry,consentRepository, dataFlowRequestPublisher, hiuProperties));
+                consentRepository,gatewayServiceClient, centralRegistry,
+                dataFlowRequestPublisher, hiuProperties, gatewayServiceProperties, gatewayResponseCache));
         consentTasks.put(REVOKED, new RevokedConsentTask(consentRepository, healthInformationPublisher));
         consentTasks.put(EXPIRED, new ExpiredConsentTask(consentRepository, dataFlowDeletePublisher));
+        consentTasks.put(DENIED, new DeniedConsentTask(consentRepository));
         this.gatewayResponseCache = CacheBuilder
-                    .newBuilder()
-                    .maximumSize(50)
-                    .expireAfterWrite(1, TimeUnit.HOURS)
-                    .build(new CacheLoader<>() {
-                        public Optional<Object> load(String key) {
-                            return Optional.empty();
-                        }
-                    });
-    }
-
-    private Flux<Void> handleConsentArtefactNotification(ConsentNotification notification, LocalDateTime timestamp) {
-        return Flux.fromIterable(notification.getConsentArtefacts())
-                .flatMap(consentArtefactReference -> {
-                    ConsentTask consentTask = consentTasks.get(notification.getStatus());
-                    if (consentTask != null) {
-                        return consentTask.perform(consentArtefactReference, notification.getConsentRequestId(), timestamp);
+                .newBuilder()
+                .maximumSize(50)
+                .expireAfterWrite(1, TimeUnit.HOURS)
+                .build(new CacheLoader<>() {
```

Comment:
This Cache can be injected as well. But that can be done as a different PR, once we have integrated

Reply:
Yeah, Also the ConsentTasks can be injected.

Reply:
ok